dbt/models/marts/suivi_complet_etps_conventionnes.py

```diff
@@ -73,6 +67,7 @@ def fill_with_zeroes(x, year):
                             "nom_region_af",
                             "nom_region_structure",
                             "siret_structure",
+                            "type_structure_emplois",
```

Comment:
ce que je racontais à propos de récupérer les titres directement depuis la dataframe ? sinon tu vas devoir MAJ ce modèle à chaque fois que les modèles sous-jacents changent ?

Reply:
Pardon j'avais mal compris, c'est en effet plutot faisable

Reply:
J'ai essayé de faire ça 

```
def get_zero_etp(df):
    def fill_with_zeroes(x, year):
        return x.reindex(
            pd.period_range(str(year) + "-01-01", str(year) + "-12-31", freq="M", name="af_date_fin_effet_v2"),
            fill_value=0,
        )

    current_year = datetime.date.today().year
    three_last_years = [current_year - 2, current_year - 1, current_year]
    df_etp_null = []
    columns_to_exclude = [
        "effectif_mensuel_conventionné",
        "effectif_annuel_conventionné",
        "nb_brsa_cible_mensuel",
        "nb_brsa_cible_annuel",
        "af_date_fin_effet_v2",
    ]
    columns_group_by = [col for col in df.columns if col not in columns_to_exclude]

    # Not all structures have a funding for the whole year therefore we create
    # a second table with the non contracted months
    # we set all months = 0 in order to compare with df_replicate
    # and remove the duplicates where df_replicate[etp] != 0
    for year in three_last_years:
        df_temp = df[df["annee_af"] == year]
        df_temp["af_date_debut_effet_v2"] = pd.to_datetime(df_temp["af_date_debut_effet_v2"])
        df_temp["af_date_fin_effet_v2"] = pd.to_datetime(df_temp["af_date_fin_effet_v2"])

        for i in df_temp.index:
            if df_temp["effectif_mensuel_conventionné"].loc[i] != 0:
                df_temp = (
                    df_temp.set_index("af_date_fin_effet_v2")
                    .groupby(columns_group_by)[
                        [
                            "effectif_mensuel_conventionné",
                            "effectif_annuel_conventionné",
                            "nb_brsa_cible_mensuel",
                            "nb_brsa_cible_annuel",
                        ]
                    ]
                    .apply(lambda x: fill_with_zeroes(x, year))
                    .reset_index()
                )

        df_temp["af_date_fin_effet_v2"] = df_temp["af_date_fin_effet_v2"].dt.to_timestamp(how="end").dt.normalize()
        df_temp = df_temp[df.columns[:-1]]
        df_etp_null.append(df_temp)

    return df_etp_null
```

et je me retrouve avec cette erreur `unhashable type: 'DatetimeIndex'` que je n'arrive pas à résoudre

Reply:
`df.columns` n'est pas une simple liste de chaînes de caractères. `df` est la représentation d'une table SQL contenant potentiellement des millions de lignes, chaque colonne ayant un type, un nom, des métadonnées, etc. Bref ce n'est pas une simple liste et c'est pour ça qu'il te dit qu'il n'est pas content.

Si spécifiquement tu veux les noms de colonnes, utilise `df.columns.values.tolist()`

Reply:
J'avais essayé et aussi bien avec `[col for col in df.columns if col not in columns_to_exclude]` ou `[col for col in df.columns.values.tolist() if col not in columns_to_exclude]`

j'ai `columns_group_by` qui est une liste

Reply:
Eh bien...... mystère, du coup je vais devoir essayer par moi-meme.

Reply:
Ben chez moi, ça marche dans les deux cas.

Reply:
j'ai la même erreur que yannick dans les deux cas